src/utils/paths.js

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (c) 2017-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+/* @flow */
+
+import { AbstractValue, ConcreteValue, Value } from "../values/index.js";
+import invariant from "../invariant.js";
+import simplifyAbstractValue from "../utils/simplifier.js";
+
+export function withPathCondition<T>(condition: AbstractValue, evaluate: () => T): T {
+  let realm = condition.$Realm;
+  let savedPath = realm.pathConditions;
+  realm.pathConditions = [];
+  try {
+    pushPathCondition(condition);
+    pushRefinedConditions(savedPath);
+    return evaluate();
+  } finally {
+    realm.pathConditions = savedPath;
+  }
+}
+
+export function withInversePathCondition<T>(condition: AbstractValue, evaluate: () => T): T {
+  let realm = condition.$Realm;
+  let savedPath = realm.pathConditions;
+  realm.pathConditions = [];
+  try {
+    pushInversePathCondition(condition);
+    pushRefinedConditions(savedPath);
+    return evaluate();
+  } finally {
+    realm.pathConditions = savedPath;
+  }
+}
+
+function pushPathCondition(condition: Value) {
+  if (condition instanceof ConcreteValue) return;
+  if (!condition.mightNotBeTrue()) return;
+  invariant(condition instanceof AbstractValue);
+  let realm = condition.$Realm;
+  if (condition.kind === "&&") {
+    let left = condition.args[0];
+    let right = condition.args[1];
+    invariant(left instanceof AbstractValue);
+    invariant(right instanceof AbstractValue);
+    pushPathCondition(left);
+    pushPathCondition(right);
+  } else {
+    realm.pathConditions.push(condition);
+  }
+}
+
+function pushInversePathCondition(condition: Value) {
+  if (condition instanceof ConcreteValue) return;
+  invariant(condition instanceof AbstractValue);
+  if (condition.kind === "||") {
+    let left = condition.args[0];
+    let right = condition.args[1];
+    invariant(left instanceof AbstractValue);
+    invariant(right instanceof AbstractValue);
+    pushInversePathCondition(left);
+    pushInversePathCondition(right);
+  } else {
+    let realm = condition.$Realm;
+    let inverseCondition = AbstractValue.createFromUnaryOp(realm, "!", condition);
+    pushPathCondition(inverseCondition);
+    let simpliedInverseCondition = simplifyAbstractValue(realm, inverseCondition);
+    if (simpliedInverseCondition instanceof AbstractValue && simpliedInverseCondition !== inverseCondition)
```

Comment:
It's strange to see this one-off case. Maybe this is to solve a little issue at hand, but I don't want to see our reasoning infrastructure being randomized.

Reply:
I'll remove the check for AbstractValue; that is no longer needed. The equality check is there to avoid duplication in the case where an expression failed to actually simplify.

Reply:
What I meant with the one-off is that only here you also add the simplified path condition in addition to a non-simplified path condition. All the other pieces that get pushed might not be simplified. Overall, I am confused whether pushed path conditions are simplified or not. I don't see enforcement or comments, just this one special treatment.